massa-grpc/src/tests/public.rs

```diff
@@ -0,0 +1,1707 @@
+        warn_announced_version_ratio: Ratio::new_raw(30, 100),
+    };
+
+    let mip_store = MipStore::try_from(([], mip_stats_config)).unwrap();
+
+    let service = MassaPublicGrpc {
+        consensus_controller: Box::new(consensus_controller),
+        consensus_channels,
+        execution_controller: execution_ctrl.0.clone(),
+        execution_channels: ExecutionChannels {
+            slot_execution_output_sender,
+        },
+        pool_channels: PoolChannels {
+            endorsement_sender,
+            operation_sender,
+            selector: selector_ctrl.0.clone(),
+            execution_controller: execution_ctrl.0.clone(),
+        },
+        pool_controller: pool_ctrl.0,
+        protocol_controller: Box::new(MockProtocolController::new()),
+        protocol_config: ProtocolConfig::default(),
+        selector_controller: selector_ctrl.0,
+        storage: shared_storage,
+        grpc_config: grpc_config.clone(),
+        version: *VERSION,
+        node_id: NodeId::new(keypair.get_public_key()),
+        keypair_factory: KeyPairFactory {
+            mip_store: mip_store.clone(),
+        },
+    };
+
+    let stop_handle = service.serve(&grpc_config).await.unwrap();
+    // std::thread::sleep(Duration::from_millis(100));
+
+    // start grpc client and connect to the server
+    let channel = tonic::transport::Channel::from_static("grpc://localhost:8888")
+        .connect()
+        .await
+        .unwrap();
+
+    let mut res = PublicServiceClient::new(channel);
+
+    let s = res.get_status(GetStatusRequest {}).await.unwrap();
+    dbg!(s);
+    stop_handle.stop();
+    std::thread::sleep(Duration::from_millis(1000));
+}
+
+#[tokio::test]
+#[serial]
+async fn get_status() {
+    let addr: SocketAddr = "[::]:4001".parse().unwrap();
+    let mut public_server = grpc_public_service(&addr);
+
+    let mut exec_ctrl = MockExecutionCtrl::new();
+    exec_ctrl
+        .expect_query_state()
+        .returning(|_| massa_execution_exports::ExecutionQueryResponse {
+            responses: vec![],
+            candidate_cursor: massa_models::slot::Slot::new(0, 2),
+            final_cursor: Slot::new(0, 0),
+            final_state_fingerprint: massa_hash::Hash::compute_from(&Vec::new()),
+        });
+
+    public_server.execution_controller = Box::new(exec_ctrl);
+
+    let config = public_server.grpc_config.clone();
+    let stop_handle = public_server.serve(&config).await.unwrap();
+    // start grpc client and connect to the server
+    let mut public_client = PublicServiceClient::connect(format!(
+        "grpc://localhost:{}",
+        addr.to_string().split(':').into_iter().last().unwrap()
+    ))
+    .await
+    .unwrap();
+    let response = public_client.get_status(GetStatusRequest {}).await.unwrap();
+    let result = response.into_inner();
+    assert_eq!(result.status.unwrap().version, *VERSION.to_string());
+
+    stop_handle.stop();
+    std::thread::sleep(Duration::from_millis(1000));
```

Comment:
@sydhds will not validate this peace of code... We should find a workaround. A helper function which shutdown a server properly.

Reply:
it can be removed